tsdb/index/tsi1/cache.go

```diff
@@ -198,3 +217,87 @@ type seriesIDCacheElement struct {
 	value       string
 	SeriesIDSet *tsdb.SeriesIDSet
 }
+
+// cacheTracker is the per-cache interface used to interact with
+// a global cache stats collector
+type cacheTracker struct {
+	// metrics is a reference to the global collector of TSI cache stats
+	metrics *tsdb.TSICacheMetrics
+	// labels is the specific combo of labels a cache must use to identify
+	// itself within the global stats
+	labels prometheus.Labels
+}
+
+func newCacheTracker(metrics *tsdb.TSICacheMetrics, defaultLabels prometheus.Labels) *cacheTracker {
+	return &cacheTracker{metrics: metrics, labels: defaultLabels}
+}
+
+// Labels returns a copy of labels for use with index cache metrics.
+func (t *cacheTracker) Labels() prometheus.Labels {
+	if t == nil {
+		return prometheus.Labels{}
+	}
+
+	l := make(map[string]string, len(t.labels))
+	for k, v := range t.labels {
+		l[k] = v
+	}
+	return l
+}
+
+func (t *cacheTracker) SetSize(sz uint64) {
+	if t == nil {
+		return
+	}
+
+	labels := t.Labels()
+	t.metrics.Size.With(labels).Set(float64(sz))
```

Comment:
For other prometheus noobs like me, I just found a great example/walkthrough of different ways a program could set up / register metrics for use across multiple "worker" components: https://github.com/prometheus/client_golang/blob/master/prometheus/examples_test.go#L160

One idea that jumps out at me is: instead of calling `With` every time the code wants to interact with a metric collector, we could just call it once and have the tracker structs keep a reference to the result.